pkg/client/optimizely_config.go

```diff
@@ -0,0 +1,163 @@
+		}
+	}
+	return variableByIDMap
+}
+
+func getExperimentVariablesMap(features []entities.Feature) (experimentVariable map[string]map[string]OptimizelyVariable) {
+	experimentVariable = map[string]map[string]OptimizelyVariable{}
+	for _, feature := range features {
+
+		var optimizelyVariableMap = map[string]OptimizelyVariable{}
+		for _, variable := range feature.VariableMap {
+			optimizelyVariableMap[variable.Key] = OptimizelyVariable{Key: variable.Key, ID: variable.ID, Value: variable.DefaultValue, Type: string(variable.Type)}
+
+		}
+		for _, experiment := range feature.FeatureExperiments {
+			experimentVariable[experiment.Key] = optimizelyVariableMap
+		}
+	}
+	return experimentVariable
+}
+
+func getExperimentMap(features []entities.Feature, experiments []entities.Experiment, variableByIDMap map[string]entities.Variable) (optlyExperimentMap map[string]OptimizelyExperiment) {
+
+	optlyExperimentMap = map[string]OptimizelyExperiment{}
+	experimentVariablesMap := getExperimentVariablesMap(features)
+
+	for _, experiment := range experiments {
+
+		var optlyVariationsMap = map[string]OptimizelyVariation{}
+		for _, variation := range experiment.Variations {
+			var optlyVariablesMap = map[string]OptimizelyVariable{}
+
+			if variableMap, ok := experimentVariablesMap[experiment.Key]; ok {
+				for index, element := range variableMap { // copy by value
+					optlyVariablesMap[index] = element
+				}
+			}
+
+			for _, variable := range variation.Variables {
+				if experiment.IsFeatureExperiment && variation.FeatureEnabled {
+					if convertedVariable, ok := variableByIDMap[variable.ID]; ok {
+						optlyVariable := OptimizelyVariable{Key: convertedVariable.Key, ID: convertedVariable.ID,
+							Type: string(convertedVariable.Type), Value: variable.Value}
+						optlyVariablesMap[convertedVariable.Key] = optlyVariable
+					}
+				}
+			}
+			optVariation := OptimizelyVariation{ID: variation.ID, Key: variation.Key, VariablesMap: optlyVariablesMap, FeatureEnabled: variation.FeatureEnabled}
+			optlyVariationsMap[variation.Key] = optVariation
+		}
+		optlyExperiment := OptimizelyExperiment{ID: experiment.ID, Key: experiment.Key, VariationsMap: optlyVariationsMap}
+		optlyExperimentMap[experiment.Key] = optlyExperiment
+	}
+	return optlyExperimentMap
+}
+
+func getFeatureMap(features []entities.Feature, variableByIDMap map[string]entities.Variable) (optlyFeatureMap map[string]OptimizelyFeature) {
+
+	optlyFeatureMap = map[string]OptimizelyFeature{}
+
+	for _, feature := range features {
+
+		var optlyFeatureVariablesMap = map[string]OptimizelyVariable{}
+		for _, featureVarible := range feature.VariableMap {
+			optlyVariable := OptimizelyVariable{Key: featureVarible.Key, ID: featureVarible.ID,
+				Type: string(featureVarible.Type), Value: featureVarible.DefaultValue}
+			optlyFeatureVariablesMap[featureVarible.Key] = optlyVariable
+		}
+
+		optlyExperimentMap := getExperimentMap(features, feature.FeatureExperiments, variableByIDMap)
```

Comment:
We do not need to rebuild map here. You can consider re-using experimentsMap built before and pick for each feature's experimentIds.

Reply:
addressed